### PR TITLE
Bugfix SOM, TC cost and Biodiversity Module + new Biodiversity Module realization

### PR DESCRIPTION
**Pull request by flohump**
## :bird: Purpose of this PR :bird:

- Bugfix in SOM module: "PreviousCarbonDensity" (no entry in changelog needed because not yet published)
- TAU: upper bound on TC costs deactived by default (no entry in changelog needed because not yet published)
- new additional data 4.21: afforestation starts in 1995 in CHA
- Bugfix in `44_biodiverstiy/bv_btc_mar21`: accounting for time step length
- New realization `44_biodiverstiy/bii_target` added AND set as new default in default.cfg

## :wrench: Checklist for PR creator :wrench:

- [x] Labeling pull request correctly [from the label list](https://github.com/magpiemodel/magpie/labels).

* Low risk : Simple bugfixes (missing files, updated documentation, typos) or Start/output scripts
* Medium risk : New realization / Changes to existing realization / Other changes which don't modify default.cfg
* High risk : New input files (if cfg$input is changed in default.cfg) / Modification to core model (eg. changes in equations, calculations, introduction of new sets etc.) / Other changes in default.cfg

- [x] Providing additional information based on PR label

* Low risk : No new model run needed.
* Medium risk : Default run based on the current version of the fork from which PR is made
* High risk
  * Default run from the current develop branch: `22LandConsvBetaRun3_default_rest1_all_marginal`
  * Default run based on the current version of the fork from which PR is made: `PR399_SSP2-Ref`
  * Run with BII increase based on the current version of the fork from which PR is made: `PR399_SSP2-Ref_BII`

Emissions_CO2_Land_Land_use_Change-51
Productivity_Landuse_Intensity_Indicator_Tau-57
Resources_Land_Cover_Cropland-68

Biodiversity_BII-3

- [x] :chart_with_downwards_trend: Performance loss/gain from current default behavior :chart_with_upwards_trend:
  * Current develop branch default : 0.72 h
  * This PR's default :  0.42 h

- [x] Added changes to `CHANGELOG.md`
- [x] Compilation check (model starts without compilation errors - use `gams main.gms action=c` in model folder for testing).
- [x] No hard coded numbers and cluster/country/region names.
- [x] The new code doesn't contain declared but unused parameters or variables.
- [x] Where relevant, In-code comments added including documentation comments.
- [x] Made sure that documentation created with [`goxygen`](https://github.com/pik-piam/goxygen) is okay (use `goxygen::goxygen()` for testing).
- [x] Changes to [`magpie4`](https://github.com/pik-piam/magpie4) R library for post processing of model output (ideally backward compatible).
- [x] Self-review of my own code.
- [x] For high risk runs: validation of major model indicators - Land-use, emissions, food prices, Tau.  %Delete this line in case it is not a high risk run%

## :warning: Additional notes or warnings :warning:
NA

## :rotating_light: Checklist for RSE reviewer :rotating_light:

- [ ] PR is labeled correctly.
- [ ] `CHANGELOG` is updated correctly
- [ ] No hard coded numbers and cluster/country/region names.
- [ ] No unnecessary increase in module interfaces
- [ ] All required updates in interfaces (if any) have been properly adressed in the module contracts
- [ ] In-code comments and documentation comments are satisfactory.
- [ ] model behavior/performance is satisfactory.
- [ ] Requested changes (if any) were applied correctly

## :rotating_light: Checklist for MAgPIE reviewer :rotating_light:

- [x] PR is labeled correctly.
- [x] `CHANGELOG` is updated correctly
- [x] No hard coded numbers and cluster/country/region names.
- [x] Changes to the model are scientifically sound
- [x] In-code comments and documentation comments are satisfactory.
- [x] model behavior/performance is satisfactory.
- [x] Requested changes (if any) were applied correctly

### CODE REVIEW

File: config/default.cfg [hunk 1]
```diff
@@ -968,22 +970,34 @@ cfg$gms$water_availability <- "total_water_aug13"    # def = total_water_aug13
...
+
+# * Start year of BII target (in bii_target) or biodiversity price (in bv_btc_mar21)
+cfg$gms$s44_start_year <- 2025	#def = 2025
+
+# * Options for realization `bii_target`
+# * Targeted annual change in BII imposed on current level, starting in `s44_start_year`
+# * Plausible values: 
+# * 	  -Inf: no target
+# * 		 0: no decrease
+# * 	0.0002: increase of BII to about 0.8 in 2050 and 0.85 in 2100
```
Comment (pvjeetze) [inline config/default.cfg:985]: maybe add very brief description that this is the yearly rate of increase to make more accessible.

File: modules/44_biodiversity/bii_target/equations.gms [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+*' @equations
+
+*' The Biodiversity Intactness Index (BII) is calculated at the level of 71 biomes.
+
+ q44_bii(biome44)$(sum(j2, f44_biome(j2,biome44)) > 0) .. v44_bii(biome44)
+ 					=e=
+ 					(sum((j2,potnatveg,landcover44), vm_bv(j2,landcover44,potnatveg) * f44_biome(j2,biome44)) / sum((j2,land), pcm_land(j2,land) * f44_biome(j2,biome44)));
+
+*' For each of the 71 biomes, the BII has to meet a minium level based on `s44_bii_target`.
```
Comment (pvjeetze) [inline modules/44_biodiversity/bii_target/equations.gms:16]: Also add that `s44_bii_target` is not the end target like e.g. 0.8 but the current value plus a given increase per time step for making it easier to understand

File: modules/41_area_equipped_for_irrigation/static/presolve.gms [hunk 3]
```diff
@@ -6,7 +6,7 @@
 *** |  Contact: magpie@pik-potsdam.de
 
 $ifthen "%c41_initial_irrigation_area%" == "Siebert" v41_AEI.fx(j)    = f41_irrig(j);
-$elseif "%c41_initial_irrigation_area%" == "LUH2v2"  v41_AEI.fx(j)    = f41_irrig_luh(t,j);
+$elseif "%c41_initial_irrigation_area%" == "LUH2v2"  v41_AEI.fx(j)    = f41_irrig_luh("y1995",j);
```
Comment (tscheypidi) [inline modules/41_area_equipped_for_irrigation/static/presolve.gms:9]: should that change be mentioned in the CHANGELOG as well?

File: modules/44_biodiversity/bii_target/equations.gms (hunk 2 above)
Comment (pvjeetze) [reply modules/44_biodiversity/bii_target/equations.gms:16]: Just thinking about it: to make it more intuitive a renaming of `s44_bii_target` to something like `s44_bii_increment`or `s44_bii_gain` or `s44_bii_growth`

**Review by tscheypidi — APPROVED** [review]
looks good, just one part of code where I had the feeling that it is not yet mentioned in the CHANGELOG (see my separate comment)

File: modules/41_area_equipped_for_irrigation/static/presolve.gms (hunk 3 above)
Comment (flohump) [reply modules/41_area_equipped_for_irrigation/static/presolve.gms:9]: True. That was an addtional bug, which I discovered and fixed.

File: modules/44_biodiversity/bii_target/realization.gms [hunk 4]
```diff
@@ -0,0 +1,25 @@
+*** |  (C) 2008-2021 Potsdam Institute for Climate Impact Research (PIK)
+*** |  authors, and contributors see CITATION.cff file. This file is part
+*** |  of MAgPIE and licensed under AGPL-3.0-or-later. Under Section 7 of
+*** |  AGPL-3.0, you are granted additional permissions described in the
+*** |  MAgPIE License Exception, version 1.0 (see LICENSE file).
+*** |  Contact: magpie@pik-potsdam.de
+
+*' @description In this realisation, the Biodiversity Intactness Index (BII) is 
+*' calculated at the level of 71 biomes.
```
Comment (pvjeetze) [inline modules/44_biodiversity/bii_target/realization.gms:9]: "...is calculated separately for each biome type of each biogeographic realm, which results in 71 different spatial units." Cite the Olsen et al. paper https://doi.org/10.1641/0006-3568(2001)051[0933:TEOTWA]2.0.CO;2

File: modules/44_biodiversity/bv_btc_mar21/preloop.gms [hunk 5]
```diff
@@ -6,5 +6,6 @@
 *** |  Contact: magpie@pik-potsdam.de
 
 v44_bv_weighted.l(j) = 0.3;
-m_sigmoid_interpol(p44_price_bv_loss,s44_start_year,s44_target_year,s44_start_price,s44_target_price);
+m_linear_interpol(p44_price_bv_loss,s44_start_year,s44_target_year,s44_start_price,s44_target_price);
```
Comment (pvjeetze) [inline modules/44_biodiversity/bv_btc_mar21/preloop.gms:9]: Why this change from sigmoid to linear?

**Review by pvjeetze — COMMENTED** [review]
I absolutely like these new additions to the biodiversity module! Many thanks! I have only minor suggestions that you'll find directly in the code.

File: modules/44_biodiversity/bii_target/realization.gms (hunk 4 above)
Comment (flohump) [reply modules/44_biodiversity/bii_target/realization.gms:9]: Thanks for these details.

File: modules/44_biodiversity/bv_btc_mar21/preloop.gms (hunk 5 above)
Comment (flohump) [reply modules/44_biodiversity/bv_btc_mar21/preloop.gms:9]: I noticed that the sigmoid price increase has almost no effect until the inflection point.

Comment (pvjeetze) [reply modules/44_biodiversity/bv_btc_mar21/preloop.gms:9]: 👍

File: modules/44_biodiversity/bii_target/equations.gms (hunk 2 above)
Comment (flohump) [reply modules/44_biodiversity/bii_target/equations.gms:16]: I updated the implementation, based on your suggestions. 
`s44_bii_target`is now the direct global BII target value for the target year.

File: config/default.cfg [hunk 6]
```diff
@@ -967,24 +969,61 @@ cfg$gms$water_availability <- "total_water_aug13"    # def = total_water_aug13
...
+# * Year in which the BII target value (in bii_target) or biodiversity target price (in bv_btc_mar21) is reached
+cfg$gms$s44_target_year <- 2100		#def = 2100
+
+# ** Options for realization `bii_target`
+# * This realisation allows to define a target value and year for the Biodiversity Intactness Index (BII)
+# * that is applied in each biome type of each biogeographic realm or at the global level. 
+# * The BII has been proposed as a control variable for changes in ecosystem functioning at the global scale
+# * and has shown to capture increasing levels of ecosystem degradation. Steffen et al. (2015, 10.1126/science.1259855) have proposed 
+# * a premliminary boundary of 0.9 at the biome level (safe space) but with an uncertainty range of 0.9-0.3,
+# * in which ecosystem degradation may undermine stable ecosystem functioning the supply of nature's
```
Comment (pvjeetze) [inline config/default.cfg:990]:
```suggestion
# * in which ecosystem degradation may undermine stable ecosystem functioning and the supply of nature's 
```

File: config/default.cfg [hunk 7]
```diff
@@ -967,24 +969,61 @@ cfg$gms$water_availability <- "total_water_aug13"    # def = total_water_aug13
...
+# * a premliminary boundary of 0.9 at the biome level (safe space) but with an uncertainty range of 0.9-0.3,
+# * in which ecosystem degradation may undermine stable ecosystem functioning the supply of nature's 
+# * contributions to people (NCP). With BII values below 0.3, the area's biodiversity has been depleted,
+# * causing high risk of ecosystem collapse. Estimates for the global BII range between 0.7975 in 1970 and 0.7754 
+# * in 2000. https://www.nhm.ac.uk/our-science/data/biodiversity-indicators/biodiversity-intactness-index-data
+# * In MAgPIE, the global BII value is about 0.8 in 1995 and declines to 0.795 by 2020. 
+
+# * BII target mode (binary)
+# *   options:   0 (global) or 1 (biome)
+# * In case of 0 (global), the BII target `s44_bii_target` is converted into increasing lower bounds for the BII in each of the 71 biomes with to gaol to meet the BII target at global level in the target year.
```
Comment (pvjeetze) [inline config/default.cfg:998]:
```suggestion
# * In case of 0 (global), the BII target `s44_bii_target` is converted into increasing lower bounds for the BII in each of the 71 biomes with to goal to meet the BII target at global level in the target year.
```

File: modules/44_biodiversity/bii_target/presolve.gms [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+	p44_start_value(biome44) = v44_bii.l(biome44);
+* In the global case (s44_bii_target_mode = 0), the target value is the start value scaled with the ratio of the global BII target `s44_bii_target` and global BII level in the start year
+	if(s44_bii_target_mode = 0,
+		p44_target_value(biome44) = (s44_bii_target / (sum((j,potnatveg,landcover44), vm_bv.l(j,landcover44,potnatveg)) / sum((j,land), pcm_land(j,land)))) * v44_bii.l(biome44);
+	elseif s44_bii_target_mode = 1,
+		p44_target_value(biome44) = s44_bii_target
+	);
+	
+* Linear increase of BII target values at biome level from start year to target year, and constant values thereafter.
+	p44_bii_target(t2,biome44) = p44_start_value(biome44) + ((m_year(t2)-s44_start_year) / (s44_target_year-s44_start_year)) * (p44_target_value(biome44)-p44_start_value(biome44));
```
Comment (pvjeetze) [inline modules/44_biodiversity/bii_target/presolve.gms:19]:
```suggestion
	p44_bii_target(t2,biome44) = p44_start_value(biome44) + ((m_year(t2) - s44_start_year) / (s44_target_year - s44_start_year)) * (p44_target_value(biome44) - p44_start_value(biome44));
```

**Review by pvjeetze — COMMENTED** [review]
Looks good to me, just found a couple of typos that I changed directly

**Review by pvjeetze — APPROVED** [review]
